pkg/operator/controller/externaldns/rbac_test.go

```diff
@@ -49,12 +40,17 @@ func TestEnsureExternalDNSClusterRole(t *testing.T) {
 		existingObjects []runtime.Object
 		expectedExist   bool
 		expectedRole    rbacv1.ClusterRole
+		errExpected     bool
 	}{
 		{
 			name:            "Does not exist",
 			existingObjects: []runtime.Object{},
 			expectedExist:   true,
 			expectedRole: rbacv1.ClusterRole{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "ClusterRole",
+					APIVersion: "rbac.authorization.k8s.io/v1",
+				},
```

Comment:
Out of curiosity, why is `TypeMeta` needed now?

Reply:
Good question, I didn't find the answer. It's somewhere in the difference of `reflect.DeepEqual` (used before) and `go-cmp` (used now).
I don't know why but `go-cmp` manages to compare `TypeMeta` and `ObjectMeta` embedded structs while reflect's `DeepEqual` does not.

I added TypeMeta fields to the ignore list to safe some space. I think that the type check is enough.

Reply:
Were the fields rbacv1.ClusterRole{}, "ResourceVersion", "Kind", "APIVersion"  ignored  when you used reflect.DeepEqual ?

Reply:
in the test case I did the following change

if gotExist {
                //diffOpts := cmpopts.IgnoreFields(rbacv1.ClusterRole{}, "ResourceVersion", "Kind", "APIVersion")
                //if diff := cmp.Diff(tc.expectedRole, *gotRole, diffOpts); diff != "" {
                if diff := reflect.DeepEqual(tc.expectedRole, *gotRole); diff != true {
                    t.Errorf("unexpected cluster role (-want +got):\n%v\n%v", tc.expectedRole, *gotRole)
                }
            }
        })
    }

and in rbac.go

//if diff := cmp.Diff(expectedRuleMap, currentRuleMap); diff != "" {
if diff := reflect.DeepEqual(expectedRuleMap, currentRuleMap); diff != true {
   return true, fmt.Sprintf("diff found in the policy rules: \n%v\n%v", expectedRuleMap, currentRuleMap)
}

I got this ->

rbac_test.go:176: unexpected cluster role (-want +got):
        {{ } {external-dns      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} [{[get list watch] [networking.k8s.io] [ingresses] [] []} {[get list watch] [] [endpoints services pods nodes] [] []}] nil}
        {{ClusterRole rbac.authorization.k8s.io/v1} {external-dns     1000 0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} [{[get list watch] [networking.k8s.io] [ingresses] [] []} {[get list watch] [] [endpoints services pods nodes] [] []}] nil}

in the expected {ClusterRole rbac.authorization.k8s.io/v1} is missing

that is why the reflect is failing.

Using cmp as you ignore these fields then the test will pass

Reply:
test output when I use reflect

[miheer@localhost external-dns-operator]$ 
[miheer@localhost external-dns-operator]$ GO111MODULE=on go test -v -run TestEnsureExternalDNSClusterRole ./pkg/...
?   	github.com/openshift/external-dns-operator/pkg/operator	[no test files]
?   	github.com/openshift/external-dns-operator/pkg/operator/config	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/openshift/external-dns-operator/pkg/operator/controller	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/openshift/external-dns-operator/pkg/operator/controller/credentials-secret	(cached) [no tests to run]
=== RUN   TestEnsureExternalDNSClusterRole
=== RUN   TestEnsureExternalDNSClusterRole/Does_not_exist
2021-09-28T00:45:51.495+1000	INFO	created externalDNS cluster role	{"name": "external-dns"}
    rbac_test.go:176: unexpected cluster role (-want +got):
        {{ } {external-dns      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} [{[get list watch] [networking.k8s.io] [ingresses] [] []} {[get list watch] [] [endpoints services pods nodes] [] []}] nil}
        {{ClusterRole rbac.authorization.k8s.io/v1} {external-dns     1 0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} [{[get list watch] [networking.k8s.io] [ingresses] [] []} {[get list watch] [] [endpoints services pods nodes] [] []}] nil}
=== RUN   TestEnsureExternalDNSClusterRole/Exists_and_as_expected
    rbac_test.go:176: unexpected cluster role (-want +got):
        {{ } {external-dns      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} [{[get list watch] [networking.k8s.io] [ingresses] [] []} {[get list watch] [] [endpoints services pods nodes] [] []}] nil}
        {{ClusterRole rbac.authorization.k8s.io/v1} {external-dns     999 0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} [{[get list watch] [networking.k8s.io] [ingresses] [] []} {[get list watch] [] [endpoints services pods nodes] [] []}] nil}
=== RUN   TestEnsureExternalDNSClusterRole/Exists_and_needs_to_be_updated
2021-09-28T00:45:51.496+1000	INFO	updated externalDNS cluster role	{"name": "external-dns", "reason": "diff found in the policy rules: \nmap[/endpoints:[get list watch] /nodes:[get list watch] /pods:[get list watch] /services:[get list watch] networking.k8s.io/ingresses:[get list watch]]\nmap[/endpoints:[get list watch] /pods:[get list watch] /services:[get list watch] networking.k8s.io/ingresses:[get list watch]]"}
    rbac_test.go:176: unexpected cluster role (-want +got):
        {{ } {external-dns      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} [{[get list watch] [networking.k8s.io] [ingresses] [] []} {[get list watch] [] [endpoints services pods nodes] [] []}] nil}
        {{ClusterRole rbac.authorization.k8s.io/v1} {external-dns     1000 0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} [{[get list watch] [networking.k8s.io] [ingresses] [] []} {[get list watch] [] [endpoints services pods nodes] [] []}] nil}
--- FAIL: TestEnsureExternalDNSClusterRole (0.00s)
    --- FAIL: TestEnsureExternalDNSClusterRole/Does_not_exist (0.00s)
    --- FAIL: TestEnsureExternalDNSClusterRole/Exists_and_as_expected (0.00s)
    --- FAIL: TestEnsureExternalDNSClusterRole/Exists_and_needs_to_be_updated (0.00s)
=== RUN   TestEnsureExternalDNSClusterRoleBinding
=== RUN   TestEnsureExternalDNSClusterRoleBinding/Does_not_exist
2021-09-28T00:45:51.496+1000	INFO	created externalDNS cluster role binding	{"name": "external-dns-test"}
=== RUN   TestEnsureExternalDNSClusterRoleBinding/Exists_and_as_expected
=== RUN   TestEnsureExternalDNSClusterRoleBinding/Exists_and_needs_to_be_updated
2021-09-28T00:45:51.496+1000	INFO	updated externalDNS cluster role binding	{"name": "external-dns-test", "reason": "following fields changed: role-name,subject-name,subject-namespace"}
--- PASS: TestEnsureExternalDNSClusterRoleBinding (0.00s)
    --- PASS: TestEnsureExternalDNSClusterRoleBinding/Does_not_exist (0.00s)
    --- PASS: TestEnsureExternalDNSClusterRoleBinding/Exists_and_as_expected (0.00s)
    --- PASS: TestEnsureExternalDNSClusterRoleBinding/Exists_and_needs_to_be_updated (0.00s)
FAIL
FAIL	github.com/openshift/external-dns-operator/pkg/operator/controller/externaldns	0.012s
?   	github.com/openshift/external-dns-operator/pkg/operator/controller/externaldns/test	[no test files]
FAIL
[miheer@localhost external-dns-operator]$

Reply:
Thanks, @miheer , your test was right. It's my previous code (before `go-cmp` change) which was not doing the right condition (`!` was missing before `DeepEqual` call).
That's why `TypeMeta` had to be added in this PR - because I not only migrated to `go-cmp` but also corrected the `if` condition.